### ISSUE TITLE
HtmlUnit / MockMvc / Boot integration

### ISSUE THREAD

**Issue by rwinch** [spring-projects/spring-boot#5263]
Similar to #2708 it would be nice to provide an easy way to provide HtmlUnit / MockMvc / Boot integration.

* [Create WebClient](http://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#spring-mvc-test-server-htmlunit-mah-setup)
* [Create WebDriver](http://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#spring-mvc-test-server-htmlunit-webdriver-setup)

Since #2708 provides an easy way to create MockMvc, perhaps that is the answer to this. However, it might be nice to be able to easily inject a `WebDriver` or `WebClient` instance similar to `@MockBean`.

**Comment by philwebb** [comment]
This fits nicely with the testing theme of 1.4

**Comment by wilkinsona** [comment]
@robwinch I can see how this might work for the simple case:

```
webClient = MockMvcWebClientBuilder
	.webAppContextSetup(context)
	.build();
```

The benefits become less clear when more complex configuration is involved:

```
webClient = MockMvcWebClientBuilder
		.mockMvcSetup(mockMvc)
		// for illustration only - defaults to ""
		.contextPath("")
		// By default MockMvc is used for localhost only;
		// the following will use MockMvc for example.com and example.org as well
		.useMockMvcForHosts("example.com","example.org")
		.build();
```

I'm not sure how Boot could figure out any of the more complex configuration options, and configuring them via a custom annotation, the environment, etc feels like it might be worse than using the existing builder API.

What did you have in mind?

**Comment by rwinch** [comment]
@wilkinsona I agree with your analysis. However, I think there are many times this is quite useful.

This limitation is also quite true of https://github.com/spring-projects/spring-boot/issues/2708#issuecomment-189308710

Perhaps an option is to allow `@Autowire` of the `MockMvcWebClientBuilder` as well. Or perhaps we just say at that point you are responsible for building it and we are making the 90%+ use case easy.

Thoughts?

PS: Sorry for my inconsistent social handles. I'm @rwinch on GithHub :)

**Comment by wilkinsona** [comment]
@rwinch (sorry `@robwinch`) Thanks. Glad to see we're on the same page.

> This limitation is also quite true of #2708 (comment)
> Perhaps an option is to allow @Autowire of the MockMvcWebClientBuilder as well.

That's actually what the title of #2708 suggests, but I think there's value in injecting a `MockMvc` instance too.

I can see three different "stages", ordered from most to least involvement of Boot:

1. Injection of `MockMvc` or `WebClient` built with sensible auto-configured defaults
2. Injection of `MockMvcBuilder` or `MockMvcWebClientBuilder` with sensible auto-configured defaults that can be further customised
3. User needs more control and does their own thing entirely

**Comment by rwinch** [comment]
@wilkinsona The levels of control you listed seems quite reasonable to me.

You might consider adding [WebDriver](http://docs.spring.io/spring/docs/current/spring-framework-reference/htmlsingle/#spring-mvc-test-server-htmlunit-webdriver-setup) as well. This confusion is likely my fault as I provided two links to `WebClient` in the initial description. I just updated it to provide the correct link now.

**Comment by philwebb** [comment]
Fixed in commit bbc91cc03f5df31c1985811d07abe4f9a906355e

**Comment by php-coder** [comment]
This (and other 5 closed issues) isn't in `in-progress` right?
